x-pack/plugin/esql/compute/src/main/java/org/elasticsearch/compute/aggregation/blockhash/TimeSeriesBlockHash.java

```diff
@@ -0,0 +1,155 @@
+        AggregatorMode mode,
+        int tsHashChannel,
+        int timestampChannel,
+        Rounding rounding,
+        DriverContext driverContext
+    ) {
+        super(driverContext.blockFactory());
+        this.mode = mode;
+        this.tsHashChannel = tsHashChannel;
+        this.timestampChannel = timestampChannel;
+        this.preparedRounding = rounding != null ? rounding.prepareForUnknown() : null;
+        this.tsidHashes = new BytesRefHash(1, blockFactory.bigArrays());
+        this.intervalHash = new LongLongHash(1, blockFactory.bigArrays());
+    }
+
+    @Override
+    public void close() {
+        Releasables.close(tsidHashes, intervalHash);
+    }
+
+    @Override
+    public void add(Page page, GroupingAggregatorFunction.AddInput addInput) {
+        BytesRefBlock tsHashBlock = page.getBlock(tsHashChannel);
+        BytesRefVector tsHashVector = Objects.requireNonNull(tsHashBlock.asVector());
+        try (var ordsBuilder = blockFactory.newIntVectorBuilder(tsHashVector.getPositionCount())) {
+            LongBlock timestampIntervalBlock = page.getBlock(timestampChannel);
+            BytesRef spare = new BytesRef();
+            if (preparedRounding != null) {
+                for (int i = 0; i < tsHashVector.getPositionCount(); i++) {
+                    BytesRef tsHash = tsHashVector.getBytesRef(i, spare);
+                    long timestampInterval = preparedRounding.round(timestampIntervalBlock.getLong(i));
+                    if (tsHash.equals(previousTsidHash) == false || timestampInterval != previousTimestamp) {
+                        long tsidOrdinal = tsidHashes.add(tsHash);
+                        if (tsidOrdinal < 0) {
+                            tsidOrdinal = -1 - tsidOrdinal;
+                        }
+                        groupOrdinal = intervalHash.add(tsidOrdinal, timestampInterval);
+                        if (groupOrdinal < 0) {
+                            groupOrdinal = -1 - groupOrdinal;
+                        }
+                        previousTsidHash = BytesRef.deepCopyOf(tsHash);
+                        previousTimestamp = timestampInterval;
+                    }
+                    ordsBuilder.appendInt(Math.toIntExact(groupOrdinal));
+                }
+            } else {
+                for (int i = 0; i < tsHashVector.getPositionCount(); i++) {
+                    BytesRef tsHash = tsHashVector.getBytesRef(i, spare);
+                    if (tsHash.equals(previousTsidHash) == false) {
+                        groupOrdinal = tsidHashes.add(tsHash);
+                        if (groupOrdinal < 0) {
+                            groupOrdinal = -1 - groupOrdinal;
+                        }
+                        previousTsidHash = BytesRef.deepCopyOf(tsHash);
+
+                        // keep last time stamp around:
+                        long timestamp = timestampIntervalBlock.getLong(i);
+                        intervalHash.add(groupOrdinal, timestamp);
```

Comment:
I think we miss hashing the rounded timestamps and use it as the `final` groupOrd?

Reply:
I this case no interval rounding is provided (and `preparedRounding` is null), so I think we can just use the timestamp directly? This is the case when no interval is provided in the es|ql query.

Reply:
> I this case no interval rounding is provided (and preparedRounding is null), so I think we can just use the timestamp directly? This is the case when no interval is provided in the es|ql query.

@martijnvg I think we might be misunderstanding each other here. In ESQL, there are two modes of grouping aggregation: partial and final (we also have single and intermediate). I think we should specify the rounding for the partial mode, where the raw input includes timestamps and tsids. The output keys for the partial aggregation are the rounded timestamp and tsid.

For the, since the input are the output keys of the partial aggregation, the timestamps are already rounded, so there's no need to specify the rounding. However, we still need to handle the grouping. Using the rounded timestamp directly can lead to a large number of groupings, with many of them being empty.

As for the rounding, we can have an extra eval operator running outside the aggregation instead of inside the block hash, but let's do it in a follow-up.

We can sync up tomorrow to clarify this?

Reply:
Yes, let's sync about this. I need to think a little bit more about this.

Reply:
I pushed: 92855af0c78e540c5d269845b975cd11b681d047

The time series grouping code is now much more simplified.
I moved the date rounding to the time series source operator instead of its own dedicated operator.

Let me know if that makes sense. I also did make use of looking up the min/max timestamp, so that a more optimized date rounding can be used.